File: frontend-updated/frontend-updated/src/services/product.service.ts
import api from "../api/client";
import { API_ENDPOINTS } from "../api/endpoints";
import { PageResponse } from "../types/api";
import { Product, Genre, Language, Subcategory, ProductRequest, MediaType } from "../types/product";

export interface BrowseParams {
  subcategoryId?: number;
  genreId?: number;
  languageId?: number;
  isRentable?: boolean;
  mediaType?: MediaType;
  keyword?: string;
  // Overlays a ProductTranslation's title/description onto each result if one
  // exists for this language - see ProductServiceImpl.overlayTranslation on
  // the backend. Distinct from `languageId` above, which FILTERS the catalog
  // by a product's own base language rather than translating display text.
  displayLanguageId?: number;
  page?: number;
  size?: number;
}

export const browseProducts = async ({
  subcategoryId,
  genreId,
  languageId,
  isRentable,
  mediaType,
  keyword,
  displayLanguageId,
  page = 0,
  size = 20,
}: BrowseParams = {}): Promise<PageResponse<Product>> => {
  const res = await api.get(API_ENDPOINTS.PRODUCTS.BASE, {
    params: { subcategoryId, genreId, languageId, isRentable, mediaType, keyword, displayLanguageId, page, size },
  });
  return res.data;
};

export const getProductById = async (productId: number, displayLanguageId?: number): Promise<Product> => {
  const res = await api.get(`${API_ENDPOINTS.PRODUCTS.BASE}/${productId}`, {
    params: displayLanguageId ? { displayLanguageId } : {},
  });
  return res.data;
};

export const getAllGenres = async (subcategoryId?: number): Promise<Genre[]> => {
  const res = await api.get(API_ENDPOINTS.PRODUCTS.GENRES, {
    params: subcategoryId ? { subcategoryId } : {},
  });
  return res.data;
};

export const getAllLanguages = async (): Promise<Language[]> => {
  const res = await api.get(API_ENDPOINTS.PRODUCTS.LANGUAGES);
  return res.data;
};

export const getAllSubcategories = async (categoryId?: number): Promise<Subcategory[]> => {
  const res = await api.get(API_ENDPOINTS.PRODUCTS.SUBCATEGORIES, {
    params: categoryId ? { categoryId } : {},
  });
  return res.data;
};

export const createProduct = async (payload: ProductRequest): Promise<Product> => {
  const res = await api.post(API_ENDPOINTS.PRODUCTS.BASE, payload);
  return res.data;
};

export const updateProduct = async (
  productId: number,
  payload: ProductRequest
): Promise<Product> => {
  const res = await api.put(`${API_ENDPOINTS.PRODUCTS.BASE}/${productId}`, payload);
  return res.data;
};

File: backend v3.0.1/src/main/java/com/bookworm/backend/service/ProductService.java
package com.bookworm.backend.service;

import com.bookworm.backend.dto.request.ProductRequest;
import com.bookworm.backend.dto.response.ProductResponse;
import com.bookworm.backend.entity.Product;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface ProductService {
    /**
     * mediaType filters the catalog by BOOK/AUDIOBOOK/VIDEO_COURSE/PODCAST
     * (null = all types). displayLanguageId, if provided and a
     * ProductTranslation exists for a given product in that language,
     * overlays the translated title/shortDescription/description onto the
     * response in place of the product's base-language text - the
     * underlying Product row (price, availability, etc.) is unaffected.
     */
    Page<ProductResponse> browse(Long subcategoryId, Long genreId, Long languageId, Boolean isRentable,
                                  Product.MediaType mediaType, String keyword, Long displayLanguageId, Pageable pageable);
    ProductResponse getById(Long productId, Long displayLanguageId);
    ProductResponse create(ProductRequest request);
    ProductResponse update(Long productId, ProductRequest request);
    void delete(Long productId);
}
